[Clang] [C++26] Expansion Statements (Part 2: Parsing and Parser Tests)

Sirraide opened the pull request:
This patch implements parsing of expansion statements and expansion initialiser lists. I’ve also had to implement sema for `CXXExpansionStmtDecl` because parsing doesn’t really work without that node.

Sirraide committed:
[Clang] [C++26] Expansion Statements (Part 2)

Sirraide committed:
[Clang] [C++26] Expansion Statements (Part 2)

erichkeane commented on clang/include/clang/Parse/Parser.h:
```diff
@@ -4186,7 +4188,8 @@ class Parser : public CodeCompletionHandler {
                                llvm::function_ref<void()>(),
-                           bool FailImmediatelyOnInvalidExpr = false);
+                           bool FailImmediatelyOnInvalidExpr = false,
+                           bool StopAtRBraceAfterComma = false);
```
This seems subtle, can we comment this?

Sirraide commented on clang/test/Parser/cxx2c-expansion-statements.cpp:
```diff
@@ -0,0 +1,63 @@
+  template for (3 : "error") // expected-error {{expansion statement declaration must declare a variable}} \
+                                expected-error {{expansion statement must be a range-based for loop}} expected-error {{TODO (expansion statements)}}
+    ;
+  template while (true) {} // expected-error {{expected '<' after 'template'}}
```
Not yet as I’m still looking into something else; I was just replying to the comments as I was adding things because I’d probably forget about some of them if I waited until after I publish everything...

zwuis commented:
P1306R5 allow only compound statement to be the body of an expansion statement. I think we need a warning.

https://github.com/cplusplus/papers/issues/156#issuecomment-2991034402

erichkeane commented on clang/lib/Parse/ParseStmt.cpp:
```diff
@@ -261,6 +261,36 @@ StmtResult Parser::ParseStatementOrDeclarationAfterAttributes(
+    if (NextToken().isOneOf(tok::kw_while, tok::kw_do)) {
+      bool IsWhile = NextToken().is(tok::kw_while);
+      Diag(Tok.getLocation(), diag::err_expansion_stmt_invalid_kw)
+          << (IsWhile ? "while" : "do");
```
SGTM!

cor3ntin commented on clang/include/clang/Basic/DiagnosticParseKinds.td:
```diff
@@ -450,6 +455,14 @@ def err_unspecified_size_with_static : Error<
 def err_expected_parentheses_around_typename : Error<
   "expected parentheses around type name in %0 expression">;
+def err_expansion_stmt_requires_range : Error<
+  "expansion statement must be a range-based for loop">;
```
I understand what you mean, but this is a bit of a confusing category error.
We should try to find a better phrasing (I am struggling to find one)

Sirraide replied:
Yeah, I was struggling w/ the wording here too; maybe ‘`template for` cannot be used with a C-style `for` loop’ or sth like that? 

(Yes, that error would probably not make much sense if someone were to try and combine `template for` w/ an Objective-C for-in loop, but I do hope no-one is actually going to try and do that...)

cor3ntin replied:
Either

> `expansion statement must use the syntax of a range-based for loop`
> `expected a range-based iteration`

Sirraide replied:
I like ‘must use the syntax of a range-based for loop’; I’ll switch it to that

cor3ntin commented on clang/include/clang/Basic/DiagnosticParseKinds.td:
```diff
@@ -450,6 +455,14 @@ def err_unspecified_size_with_static : Error<
+def err_expansion_stmt_requires_range : Error<
+  "expansion statement must be a range-based for loop">;
+def err_expansion_stmt_requires_cxx2c : Error<
+  "expansion statements are only supported in C++2c">;
```
But we have 

```
defm expansion_statements : CXX26Compat<
  "expansion statements are">;
``` 

above ?

Sirraide replied:
Ah, I think that was from before you requested that we not backport this; we can remove the extension warning for now

cor3ntin commented on clang/include/clang/Basic/DiagnosticParseKinds.td:
```diff
@@ -450,6 +455,14 @@ def err_unspecified_size_with_static : Error<
+  "expansion statement must be a range-based for loop">;
+def err_expansion_stmt_requires_cxx2c : Error<
+  "expansion statements are only supported in C++2c">;
+def err_for_template : Error<
+  "'for template' is invalid; use 'template for' instead">;
```
Haha

cor3ntin commented on clang/lib/Parse/ParseStmt.cpp:
```diff
@@ -261,6 +261,35 @@ StmtResult Parser::ParseStatementOrDeclarationAfterAttributes(
+    // QOI (and basically just ignore the 'template' token for error recovery).
+    if (NextToken().isOneOf(tok::kw_while, tok::kw_do)) {
+      Diag(Tok.getLocation(), diag::err_expansion_stmt_invalid_kw)
+          << NextToken().getKind();
```
I like fixing for `for template`, but I think this is too cute - lets not create more work for ourselves until there is evidence of user confusion

Sirraide replied:
Alright; I’ll remove this part then

Sirraide replied:
Actually I thought you meant we should remove the handling for `for template`; iirc @Endilll asked me to add code to handle `template do`/`template while`

Endilll replied:
Yes, I requested it a while ago.

> until there is evidence of user confusion

How are we going to measure that? Surely you don't expect people to show up on our bug tracker with this?

cor3ntin replied:
I also expect very few people, if any, to get into this situation

Endilll replied:
`constexpr if` is a mistake I made myself multiple times, so I don't see why `for template` is that hard to write.

Sirraide replied:
Yeah, I could definitely see some people trying to write `for template` given that it’s `if constexpr`

shafik replied:
Although if we do remove this diagnostic we should still have tests to verify the diagnostic w/ this specific check are sane'ish.

Sirraide replied:
I mean, from my point of view, adding this diagnostic isn’t exactly complicated, so imo we should just keep it

cor3ntin replied:
Nah. I really would like us to remove  the handling of `template do` / `template while`

cor3ntin commented on clang/lib/Parse/ParseStmt.cpp:
```diff
@@ -1884,8 +1898,55 @@ bool Parser::isForRangeIdentifier() {
+      getLangOpts().CPlusPlus23);
+
+  // P2718R0 - Lifetime extension in range-based for loops.
+  if (getLangOpts().CPlusPlus23) {
```
We did not - it breaks code if the range initializes a mutex and other such shenanigans

cor3ntin commented on clang/lib/Parse/ParseStmt.cpp:
```diff
@@ -1895,8 +1937,56 @@ bool Parser::isForRangeIdentifier() {
+  auto Ctx = Sema::ExpressionEvaluationContext::PotentiallyEvaluated;
+  if (FRI.ExpansionStmt && VarDeclSpec && VarDeclSpec->hasConstexprSpecifier())
+    Ctx = Sema::ExpressionEvaluationContext::ImmediateFunctionContext;
+
```
That seems suspicious, i would expect a constant evaluated context here, I need to think more about it (or for the standard to be more explicit)

Sirraide replied:
Yeah; Dan’s fork was using `ImmediateFunctionContext` for this; I’m not quite sure why candidly

cor3ntin replied:
@katzdm

katzdm replied:
I'm not sure whether Clang's use of "immediate function context" deviates from the Standard, but initializers of constexpr variables are in an immediate function context. Except for some funky cases, constexpr variables are "usable in constant expressions" ([[expr.const.init]/7.1](https://eel.is/c++draft/expr.const.init#7.1)); therefore, their initializers are manifestly constant-evaluated ([[expr.const.defns]/1.6](https://eel.is/c++draft/expr.const.defns#1.6)); therefore, all subexpressions of their initializers are in an immediate function context ([[expr.const.imm]/1.2](https://eel.is/c++draft/expr.const.imm#1.2)).

Anyway - I think that was my logic.

cor3ntin commented on clang/lib/Parse/ParseStmt.cpp:
```diff
@@ -1895,8 +1937,56 @@ bool Parser::isForRangeIdentifier() {
+    // The expansion-initializer is not in a dependent context and should
+    // thus be parsed in the parent context of the expansion statement.
+    assert(Actions.CurContext->isExpansionStmt());
+    Sema::ContextRAII CtxGuard(Actions, Actions.CurContext->getParent(),
+                               /*NewThis=*/false);
```
Why not use ExpansionStmtContextRAII ?

Sirraide replied:
We can use it here too sure; the main point of that helper is to _conditionally_ enter the parent context only if we’re parsing an expansion statement, so it’s mainly used in code that is shared with regular (range-based) for loops; that isn’t the case here, but we can still use it here too I suppose

cor3ntin commented on clang/lib/Sema/SemaExpand.cpp:
```diff
@@ -0,0 +1,76 @@
+  auto *TParam = NonTypeTemplateParmDecl::Create(
+      Context, Context.getTranslationUnitDecl(), TemplateKWLoc, TemplateKWLoc,
+      TemplateDepth, /*Position=*/0, ParmName, ParmTy, /*ParameterPack=*/false,
+      ParmTI);
```
Why does the parameter needs to have a name? We certainly ought not to look it up

Sirraide replied:
It probably doesn’t; I’ll try removing it

shafik commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -2926,10 +2926,10 @@ def note_which_delegates_to : Note<"which delegates to">;
 def err_for_range_storage_class : Error<
-  "loop variable %0 may not be declared %select{'extern'|'static'|"
-  "'__private_extern__'|'auto'|'register'|'constexpr'|'thread_local'}1">;
+  "%select{loop|expansion}0 variable %1 may not be declared %select{'extern'|'static'|"
```
It looks like you missed a few of the options, for example I don't see a test w/ `register`

Sirraide replied:
Right, I can add tests for that

shafik commented on clang/include/clang/Parse/Parser.h:
```diff
@@ -1713,11 +1713,13 @@ class Parser : public CodeCompletionHandler {
   /// Information on a C++0x for-range-initializer found while parsing a
-  /// declaration which turns out to be a for-range-declaration.
+  /// declaration which turns out to be a for-range-declaration. Also used
+  /// for C++26's expansion statements.
```
```suggestion
  /// declaration which turns out to be a for-range-declaration or a
  /// C++26's expansion statements.
```
I guess also modify the first sentence.

Sirraide replied:
That wording isn’t quite right: both expansion statements and range-based for loops contain a _for-range-declaration_. I guess nothing in the original comment is specific to range-based for loops, so I’ll just remove my original addition here

shafik commented on clang/test/Parser/cxx2c-expansion-statements.cpp:
```diff
@@ -0,0 +1,65 @@
+}
+
+void good() {
+  template for (auto y : {}); // expected-error {{TODO (expansion statements)}}
```
I guess packs come in a later PR?

Sirraide replied:
Yes, there should be tests involving packs later on

Sirraide committed:
[Clang] [C++26] Expansion Statements (Part 3: Enumerating Expansion Statements) (#169682)

This implements Sema and template instantiation for enumerating
expansion statements, as well as expansion of expansion statements in
general.
